### PR TITLE
feat(runloop): plugin:configure(configs) handler

### PR DESCRIPTION
**Pull request by bungle**
### Summary

Adds a new handler that plugins can implement:

```lua
 -- configs will be `nil` if there is no active configs for the plugin
function Plugin:configure(configs)
end
```

See the change in `acme` and `prometheus` plugins.

KAG-2672
KAG-2678
KAG-2679

### CODE REVIEW

File: kong/runloop/plugins_iterator.lua [hunk 1]
```diff
@@ -175,52 +194,50 @@ local PLUGIN_GLOBAL_KEY = build_compound_key() -- all nil
...
-            return combos[key]
-        end
+  end
+  if service_id then
+    key = build_compound_key(nil, service_id, nil)
+    if combos[key] then
+      return combos[key]
     end
-    return combos[PLUGIN_GLOBAL_KEY]
-
```
Comment (jschmid1) [inline kong/runloop/plugins_iterator.lua:223]: There seems to be no contentual change, can we revert?

File: kong/runloop/plugins_iterator.lua [hunk 2]
```diff
@@ -234,21 +251,19 @@ end
...
-    -- Filter out route, service, and consumer based on the plugin handler rules and get their ids
-  local route_id = (ctx.route and not plugin.handler.no_route) and ctx.route.id or nil
-  local service_id = (ctx.service and not plugin.handler.no_service) and ctx.service.id or nil
-  -- TODO: kong.client.get_consumer() for more consistency. This might be an exception though as we're aiming for raw speed instead of compliance.
-  local consumer_id = (ctx.authenticated_consumer and not plugin.handler.no_consumer) and ctx.authenticated_consumer.id or nil
+local function load_configuration_through_combos(ctx, combos)
+  -- Filter out route, service, and consumer based on the plugin handler rules and get their ids
+  local route_id = ctx.route and ctx.route.id or nil
+  local service_id = ctx.service and ctx.service.id or nil
+  local consumer_id = ctx.authenticated_consumer and ctx.authenticated_consumer.id or nil
```
Comment (jschmid1) [inline kong/runloop/plugins_iterator.lua:260]: This field lost the `no_consumer` validation. Is this intended?

File: kong/runloop/plugins_iterator.lua [hunk 3]
```diff
@@ -454,61 +463,79 @@ function PluginsIterator.new(version)
...
-  for _, plugin in ipairs(loaded_plugins) do
-    local name = plugin.name
-    for _, data in pairs(ws) do
-      local plugins = data.plugins
-      if plugins[name] then
-        local n = plugins[0] + 1
-        plugins[n] = plugin
-        plugins[0] = n
-        plugins[name] = nil
+  if has_plugins then
```
Comment (jschmid1) [inline kong/runloop/plugins_iterator.lua:511]: What about reversing the logic to reduce multi-level nesting?

``` lua
if not has_plugins then
... abort
end

for _, plugin...

...
end
```

File: kong/runloop/plugins_iterator.lua [hunk 4]
```diff
@@ -454,61 +463,79 @@ function PluginsIterator.new(version)
...
+      local cfg = globals[name]
+      if cfg then
+        for _, phase in ipairs(NON_COLLECTING_PHASES) do
+          if plugin.handler[phase] then
+            local plugins = globals[phase]
+            local n = plugins[0] + 2
+            plugins[0] = n
+            plugins[n] = cfg
+            plugins[n - 1] = plugin
+          end
```
Comment (jschmid1) [inline kong/runloop/plugins_iterator.lua:538]: There seems to be no change?

File: kong/runloop/plugins_iterator.lua [hunk 5]
```diff
@@ -454,61 +463,79 @@ function PluginsIterator.new(version)
...
+        if plugins[name] then
+          local n = plugins[0] + 1
+          plugins[n] = plugin
+          plugins[0] = n
+          plugins[name] = nil
+        end
+
+        local handler = HANDLERS_WITH_CONFIGURE[name]
+        if handler then
+          handler[CONFIGURE_PHASE](handler, configurable[name])
```
Comment (jschmid1) [inline kong/runloop/plugins_iterator.lua:525]: As already proposed, we should guard this function from erroring (pcall)

**Review by jschmid1 — COMMENTED** [review]
great change!

File: kong/runloop/plugins_iterator.lua (hunk 1 above)
Comment (bungle) [reply kong/runloop/plugins_iterator.lua:223]: This file seems to be using mixed spacing. `4` in some cases and `2` in some. I think it is better to clean it up.

File: kong/runloop/plugins_iterator.lua (hunk 2 above)
Comment (bungle) [reply kong/runloop/plugins_iterator.lua:262]: I don't recall what I was doing here. Perhaps it was related to getting rid of global here or something. These are performance optimizations, not validations?

File: kong/runloop/plugins_iterator.lua (hunk 4 above)
Comment (bungle) [reply kong/runloop/plugins_iterator.lua:536]: The `if cfg` was moved higher.

Comment (bungle) [reply kong/runloop/plugins_iterator.lua:536]: yes, only wrapper with if has_plugins

File: kong/runloop/plugins_iterator.lua (hunk 5 above)
Comment (bungle) [reply kong/runloop/plugins_iterator.lua]: done.

File: kong/runloop/plugins_iterator.lua (hunk 3 above)
Comment (bungle) [reply kong/runloop/plugins_iterator.lua:514]: It is hardish.

**Comment by bungle** [comment]
@jschmid1 I also tried to minimize the delta between CE and EE on this, see:
https://github.com/Kong/kong-ee/pull/6795

File: kong/plugins/acme/handler.lua [hunk 6]
```diff
@@ -70,28 +70,33 @@ local domains_matcher
...
-  end, "crud", "plugins")
+function ACMEHandler:init_worker()
+  local worker_id = ngx.worker.id()
+  kong.log.info("acme renew timer started on worker ", worker_id)
+  ngx.timer.every(86400, renew)
+end
+
+
+function ACMEHandler:configure(configs)
+  CONFIG = configs and configs[1] or nil
```
Comment (jschmid1) [inline kong/plugins/acme/handler.lua:93]: Is this fine as we'll have only one ACME plugin that is being `configure`d? due to how our plugin iterator works?

File: kong/plugins/prometheus/exporter.lua [hunk 7]
```diff
@@ -489,7 +443,7 @@ local function metric_data(write_fn)
 
   -- notify the function if prometheus plugin is enabled,
   -- so that it can avoid exporting unnecessary metrics if not
-  prometheus:metric_data(write_fn, not is_prometheus_enabled())
+  prometheus:metric_data(write_fn, not IS_PROMETHEUS_ENABLED)
```
Comment (jschmid1) [inline kong/plugins/prometheus/exporter.lua:446]: Is this covered by tests? Are we sure this still works the same way?

File: kong/runloop/handler.lua [hunk 8]
```diff
@@ -719,6 +726,8 @@ do
       end
 
       if plugins_iterator then
+        plugins_iterator:configure()
```
Comment (jschmid1) [inline kong/runloop/handler.lua:729]: Lets leave some comments explaining what's/when/where things are happening.

File: kong/runloop/plugins_iterator.lua [hunk 9]
```diff
@@ -366,8 +397,10 @@ local function get_next_and_collect(ctx, i)
   return get_next_and_collect(ctx, i)
 end
 
+
 local function get_collecting_iterator(self, ctx)
   local ws = get_workspace(self, ctx)
+  ctx.plugins = get_table_for_ctx(ws)
```
Comment (jschmid1) [inline kong/runloop/plugins_iterator.lua:403]: What's the reason for this change?

File: kong/runloop/plugins_iterator.lua [hunk 10]
```diff
@@ -390,23 +422,63 @@ local function new_ws_data()
...
+    kong_global.set_namespaced_log(kong, plugin.name, ctx)
+    local start = utils.get_updated_monotonic_ms()
+    local ok, err = pcall(plugin.handler[CONFIGURE_PHASE], plugin.handler, configurable[name])
+    local elapsed = utils.get_updated_monotonic_ms() - start
+    kong_global.reset_log(kong, ctx)
+
+    if not ok then
+      kong.log.err("failed to execute plugin '", name, ":", CONFIGURE_PHASE, " (", err, ")")
+    else
+      if elapsed > 50 then
```
Comment (jschmid1) [inline kong/runloop/plugins_iterator.lua:440]: any chance that elapsed can be `nil`?

File: kong/runloop/plugins_iterator.lua [hunk 11]
```diff
@@ -390,23 +422,63 @@ local function new_ws_data()
...
+        kong.log.notice("executing plugin '", name, ":", CONFIGURE_PHASE, " took excessively long: ", elapsed, " ms")
+      end
+    end
+  end
+end
+
+
+local function create_configure(configurable)
```
Comment (jschmid1) [inline kong/runloop/plugins_iterator.lua:448]: This doesn't do anything when either

- control plane
- subsystem == "http" and #kong.configuration.proxy_listeners == 0
- subsystem == "stream" and #kong.configuration.stream_listeners == 0

Why? and what are the cases it returns anything? Can we reverse the logic to be more descriptive?

**Review by jschmid1 — COMMENTED** [review]
We absolutely need tests for this. What about utilizing one of our mock/test plugins and implement a configure phase that is easily verifiable?

File: kong/plugins/prometheus/exporter.lua (hunk 7 above)
Comment (jschmid1) [reply kong/plugins/prometheus/exporter.lua:446]: looks like this test is exercising the code -> https://github.com/Kong/kong/blob/master/spec/03-plugins/26-prometheus/07-optional_fields_spec.lua#L119

File: kong/runloop/plugins_iterator.lua (hunk 9 above)
Comment (jschmid1) [reply kong/runloop/plugins_iterator.lua:403]: From our discussion:  "this is an alignment with kong-ee"

File: kong/runloop/plugins_iterator.lua [hunk 12]
```diff
@@ -390,23 +422,63 @@ local function new_ws_data()
...
+  if kong.configuration.role == "control_plane"
+  or ((subsystem == "http"   and #kong.configuration.proxy_listeners == 0) or
+      (subsystem == "stream" and #kong.configuration.stream_listeners == 0))
+  then
```
Comment (jschmid1) [inline kong/runloop/plugins_iterator.lua:452]: We have this check in multiple places. We may benefit from a generic function that does these checks

File: kong/runloop/plugins_iterator.lua [hunk 13]
```diff
@@ -390,23 +422,63 @@ local function new_ws_data()
...
+    return function() end
+  end
+
+  return function(self, ctx)
+    configure(configurable, ctx)
+    self.configure = nil
+    configurable = nil
+  end
```
Comment (jschmid1) [inline kong/runloop/plugins_iterator.lua:460]: This is a bit confusion. This apparently self-destructs when it's called.

File: kong/plugins/acme/handler.lua (hunk 6 above)
Comment (jschmid1) [reply kong/plugins/acme/handler.lua:93]: Answer: yes

File: kong/runloop/plugins_iterator.lua (hunk 13 above)
Comment (bungle) [reply kong/runloop/plugins_iterator.lua]: Added note there.

File: kong/runloop/plugins_iterator.lua (hunk 12 above)
Comment (bungle) [reply kong/runloop/plugins_iterator.lua:457]: I agree. Let's do it in other PR.

File: kong/runloop/plugins_iterator.lua (hunk 11 above)
Comment (bungle) [reply kong/runloop/plugins_iterator.lua:448]: I added comment and TODO on it.

File: kong/runloop/plugins_iterator.lua (hunk 10 above)
Comment (bungle) [reply kong/runloop/plugins_iterator.lua:440]: I don't think there is chance for that. We already use `get_monotonic_ms` and `get_updated_monotonic_ms` where we blindly trust that it works.

File: kong/runloop/handler.lua (hunk 8 above)
Comment (bungle) [reply kong/runloop/handler.lua:734]: Comment added.

File: kong/plugins/prometheus/exporter.lua (hunk 7 above)
Comment (bungle) [reply kong/plugins/prometheus/exporter.lua:446]: Correct.

**Review by jschmid1 — APPROVED** [review]
all my points have been addressed. approved granted the tests pass
